core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java

```diff
@@ -97,6 +101,53 @@ public void testRandomClusterStateUpdates() {
         logger.info("Final cluster state: {}", state.prettyPrint());
     }
 
+    public void testSimulateMasterDataDirectoryWiped() {
```

Comment:
I feel that the test method name does not really reflect what is happening in the test. Should it be something like "testJoiningNewClusterRemovesExistingInMemoryIndexStructures"?
Feel free to add some comments as well to what is actually tested here.

Last, I think that we don't need to simulate a whole cluster to test the behavior here. Checking one single node is enough.